logging: Replace LogError and LogWarning with LogAlert

ryanofsky opened the pull request:
Replace `LogError` and `LogWarning` with `LogAlert` to make it easier to choose the correct logging levels to use and avoid log spam. 

This PR implements an idea ajtowns came up with in [#30347 (comment)](https://github.com/bitcoin/bitcoin/pull/30347#discussion_r1659096893) and it reduces the decision tree for chosing log levels to:

```mermaid
flowchart TD
    start-->|No|dbg
    start{{Is this critical information?}}-->|Yes|crit

    dbg-->|No|logtrace(Trace + Category)
    dbg{{Is this moderate volume?}}-->|Yes|logdebug(Debug + Category)

    crit{{Is there a problem?}}-->|No|loginfo(Info)
    crit-->|Yes|logfatal(Alert)
```

ryanofsky committed:
logging: Add level::Alert and LogAlert

ryanofsky committed:
logging: Add Warning: prefixes

Add Warning: prefixes so message intent is not lost when warning level is
removed next commit.

ryanofsky commented:
> I think this makes the logging less useful, as it's less clear if the alert is a warning, an error, or fatal

I don't have a strong opinion about this change, but one thing I like about it is that it reduces the potential for log spam. I think developers will be less likely to use a level called "alert" for minor errors and warnings, and will more appropriately use the debug log level instead.

It is true that this PR gets rid of the ability to distinguish warnings and errors from informational messages and fatal errors at an API level. But this might not be a big problem if messages are clearly written and have "Error:" "Warning:" "Internal bug detected:" "Fatal error" and similar prefixes. Also practically speaking I think we are already logging most errors and warnings at the "info" and "debug" levels anyway so maybe this makes the API match current code better.

Again, though I don't have strong feeling about this change. It just seemed like it would be simple to implement and the results look pretty clean, IMO, so I think it's worth looking at and considering.

ryanofsky commented:
Rebased 179a0715cdcb3f20f71bf757c37046b0675dc8a5 -> d34b529ed104a3a7e11c7d264703e61d84b1aeb3 ([`pr/alert.1`](https://github.com/ryanofsky/bitcoin/commits/pr/alert.1) -> [`pr/alert.2`](https://github.com/ryanofsky/bitcoin/commits/pr/alert.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/alert.1-rebase..pr/alert.2)) due to silent conflict with #29625

ryanofsky committed:
scripted-diff: Replace log error and warning with alert

-BEGIN VERIFY SCRIPT-
r() { git grep -l "$1" -- 'src' ':(exclude)src/logging.h' ':(exclude)src/logging.cpp' | xargs sed -i "s/$1/$2/"; }
r LogWarning LogAlert
r LogError LogAlert
r Level::Warning Level::Alert
r Level::Error Level::Alert
sed -i -e 's/error\]/alert]/; s/warning\]/alert]/;' src/test/logging_tests.cpp
-END VERIFY SCRIPT-

ryanofsky committed:
logging: Drop unused Warning and Error levels

ryanofsky commented on src/logging.h:
re: https://github.com/bitcoin/bitcoin/pull/30364#discussion_r1675452625

> I think you can split out the 8 `/d` delete statements and three script lines for a simple last commit to just manually delete those lines

Good idea, done in latest push.

ryanofsky reviewed (commented):
Rebased d34b529ed104a3a7e11c7d264703e61d84b1aeb3 -> 4e9ff3100ae79f3c3046a3358ff16daa7cd89627 ([`pr/alert.2`](https://github.com/ryanofsky/bitcoin/commits/pr/alert.2) -> [`pr/alert.3`](https://github.com/ryanofsky/bitcoin/commits/pr/alert.3), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/alert.2-rebase..pr/alert.3)) due to conflict with #30428

maflcko reviewed (approved):
For context: The majority of the changed log messages here stem from the `error()` removal in 31be1a47675e4449f856e61beb2b4bfc228ea219 which changed `ERROR:` to `[error]`.  Many of them were not fatal error messages, but "warnings", or stuff that would not abort the program, so changing to `[alert]`, seems clearer either way.


ACK 4e9ff3100ae79f3c3046a3358ff16daa7cd89627 🔗

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: ACK 4e9ff3100ae79f3c3046a3358ff16daa7cd89627 🔗
f70Ey8roKfaWqoDaKb+Ropa/pkhackB7NyvBkH2A2MhoJj72WTdgsblLayNSYOfU5ltNUpyN5rpogffZdGyCAQ==
```

</details>

ryanofsky committed:
logging: Add noui.cpp Warning: and Error: prefixes

Suggested by mflcko in https://github.com/bitcoin/bitcoin/pull/30364#discussion_r1679920839

This should change log output in next commit from current:

  [error] Message
  [warning] Message

to:

  [alert] Error: Message
  [alert] Warning: Message

instead of:

  [alert] Message
  [alert] Message

ryanofsky commented on src/noui.cpp:
```diff
@@ -28,11 +28,11 @@ bool noui_ThreadSafeMessageBox(const bilingual_str& message, const std::string&
     case CClientUIInterface::MSG_WARNING:
         strCaption = "Warning: ";
-        if (!fSecure) LogWarning("%s\n", message.original);
+        if (!fSecure) LogAlert("%s\n", message.original);
```
re: https://github.com/bitcoin/bitcoin/pull/30364#discussion_r1679920839

> nit in [fbdd882](https://github.com/bitcoin/bitcoin/commit/fbdd8824d747c201eb01c8383d135ba08298c110): Forgot to add `Warning: ` and `Error: ` above? Can be achieved by just using `strCaption`. See also [824f472](https://github.com/bitcoin/bitcoin/commit/824f47294a309ba8e58ba8d1da0af15d8d828f43)

Good catch. Done in new commit

ryanofsky reviewed (commented):
Updated 4e9ff3100ae79f3c3046a3358ff16daa7cd89627 -> 66d35a5c1384bf579e2da8e7054be7536e7a7101 ([`pr/alert.3`](https://github.com/ryanofsky/bitcoin/commits/pr/alert.3) -> [`pr/alert.4`](https://github.com/ryanofsky/bitcoin/commits/pr/alert.4), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/alert.3..pr/alert.4)) with suggested change adding noui.cpp warning and error prefixes

@maflcko do you think you could review #30361 as well? #30361 is a documentation bugfix that I think would make this PR easier to understand, because right now the documentation that this PR is changing is not accurate.

ajtowns commented on src/node/blockstorage.cpp:
```diff
@@ -131,13 +131,13 @@ bool BlockTreeDB::LoadBlockIndexGuts(const Consensus::Params& consensusParams, s
                 pcursor->Next();
             } else {
-                LogError("%s: failed to read value\n", __func__);
+                LogAlert("%s: failed to read value\n", __func__);
```
"block index" might be clearer than "value", and could perhaps attempt to decode the block hash from the `key` and report it? Might be worth pointing out why this is a problem at a higher level, ie "block index database corrupted" or similar?

ajtowns commented on src/node/blockstorage.cpp:
```diff
@@ -131,13 +131,13 @@ bool BlockTreeDB::LoadBlockIndexGuts(const Consensus::Params& consensusParams, s
 
                 if (!CheckProofOfWork(pindexNew->GetBlockHash(), pindexNew->nBits, consensusParams)) {
-                    LogError("%s: CheckProofOfWork failed: %s\n", __func__, pindexNew->ToString());
+                    LogAlert("%s: CheckProofOfWork failed: %s\n", __func__, pindexNew->ToString());
```
Would probably be better to replace `__func__` in all these messages with something more relevant to potential users (filename, or some more general indication of what's going on). `"block index database corrupted: proof of work check failed for %s\n", pindexNew->ToString()` eg?

ajtowns commented on src/script/signingprovider.cpp:
```diff
@@ -158,7 +158,7 @@ bool FillableSigningProvider::GetKey(const CKeyID &address, CKey &keyOut) const
 {
     if (redeemScript.size() > MAX_SCRIPT_ELEMENT_SIZE) {
-        LogError("FillableSigningProvider::AddCScript(): redeemScripts > %i bytes are invalid\n", MAX_SCRIPT_ELEMENT_SIZE);
+        LogAlert("FillableSigningProvider::AddCScript(): redeemScripts > %i bytes are invalid\n", MAX_SCRIPT_ELEMENT_SIZE);
```
Should this be an `if (!Assume(redeemScript.size() <= MAX_SCRIPT_ELEMENT_SIZE)) return false;` ?

ajtowns commented on src/test/logging_tests.cpp:
```diff
@@ -140,8 +140,8 @@ BOOST_FIXTURE_TEST_CASE(logging_LogPrintMacros, LogSetup)
-    LogWarning("foo9: %s\n", "bar9");
-    LogError("foo10: %s\n", "bar10");
+    LogAlert("foo9: %s\n", "bar9");
+    LogAlert("foo10: %s\n", "bar10");
```
Not much value testing the same thing twice.

ajtowns commented on src/i2p.cpp:
```diff
@@ -148,7 +148,7 @@ bool Session::Listen(Connection& conn)
         return true;
     } catch (const std::runtime_error& e) {
-        LogPrintLevel(BCLog::I2P, BCLog::Level::Error, "Couldn't listen: %s\n", e.what());
+        LogPrintLevel(BCLog::I2P, BCLog::Level::Alert, "Couldn't listen: %s\n", e.what());
```
Touching all the `LogPrintLevel` warning/error calls without switching them to  `LogAlert` seems a shame. A separate scripted-diff prior to "Replace log error and warning..." with:

```sh
sed -i 's/LogPrintLevel(\(BCLog::[^,]*\), BCLog::Level::Error, */LogAlert(/' $files
sed -i 's/LogPrintLevel(\(BCLog::[^,]*\), BCLog::Level::Warning, */LogAlert(/' $files
```

should work, I think.

ajtowns reviewed (commented):
I think this approach makes sense, though the commit that explicitly adds "Warning" and "Error" prefixes does give me some doubts.

Either way, given we're touching a bunch of log lines, I think we should be thinking about whether they actually make sense to be alerts, so most of the following comments are in that vein.

ryanofsky commented:
I lost enthusiasm for this PR, so will close it, but tagging as up for grabs, and will happily review if someone else wants to pick it up and respond to the earlier comments, particularly the individual logging improvements suggested https://github.com/bitcoin/bitcoin/pull/30364#pullrequestreview-2182052463.

---

The thing I like about this PR is that it disentangles log message  ___priority levels___ from log message ___conditions___.

As I see it, **priority levels** can distinguish between log messages that:

- Provide information unlikely to help users, or are high volume (would call `Trace`)
- Provide information helpful for debugging user issues and are low volume (would call `Debug`)
- Provide information probably helpful in general (would call `Notify`)
- Provide information probably requiring some action (would call `Critical`)

At each of these levels, different **conditions** are possible:

- Normal conditions (would call `Info`)
- Abnormal conditions that might indicate a problem (would call `Warning`)
- Failure conditions that indicate something has gone wrong (would call `Error`)
- Impossible conditions indicating an incorrect assumption, or a bug, or hardware failure (would call `Bug`)

Examples of different priority level / condition pairs:

- Errors
  - An out of disk space error would probably be a `Critical`/`Error` message.
  - A case where saved data is missing or corrupted, but can easily be reconstructed would probably be a `Notify`/`Error` message.
  - A transient network error, or an incorrectly called RPC would probably be a `Debug`/`Error` message.
- Warnings
  - A network fork or incorrectly set system time would probably be a `Critical`/`Warning` message.
  - A valid, but obsolete or suspicious config setting would probably be a `Notify`/`Warning` message.
  - A partially successful RPC operation, or incoming data with unrecognized fields or version numbers would probably be `Debug/Warning` messages.
- Information
  - Shutdown or reindexing requests that take the node offline, or wallet operations that require generating new backups would probably be `Critical`/`Info` messages.
  - Useful, non-spammy status updates would probably be `Notify`/`Info` messages.
  - More detailed or spammy status updates would be `Debug`/`Info` or `Trace`/`Info` message.
- Bugs
  - Impossible conditions where we might use `Assert` or `std::logic_error`, and indicate something is severely wrong, like CheckBlockIndex failing, would probably be `Critical`/`Bug` messages.
   - Impossible conditions where we might use `Assume` or `std::logic_error`, but indicate a recoverable problem, like a floating point error calculating position of a progress bar, or a local bug inside of an RPC call that can be easily aborted, would probably be `Notify`/`Bug` or `Debug`/`Bug` messages.

---

In my opinion, it is important for the logging framework to be able to distinguish log message **priority levels** so important information will be seen and logs will not be filled with spam. I think having the 4 priority levels described above would be ideal (Trace Debug Notify Critical), but having the 4 levels defined in this PR would also be ok (Trace Debug Info Alert).

I don't think it's important for the logging framework know about log message **conditions**, because I think messages can just describe their own conditions clearly, and use `Error:` `Warning:` and `Bug:` prefixes to be explicit.

If people disagree and think 4 log levels is not enough, or that explicitly tracking conditions is important, probably the easiest way to support this would be expand the number of levels like [syslog](https://man7.org/linux/man-pages/man3/syslog.3.html) does, or like #30347 tried to do, so basically each priority+condition pair maps straightforwardly to some logging level.

But I don't actually care very much about what logging levels are defined, as long as their usage is [documented accurately](https://github.com/bitcoin/bitcoin/pull/30361).
